### PR TITLE
Elasticsearch: Add query building for log queries

### PR DESCRIPTION
**Pull request by ivanahuckova**
Part of https://github.com/grafana/grafana/issues/54011. Fixes failing test for log type of query that was commented out during https://github.com/grafana/grafana/pull/59578.  While on frontend [here](https://github.com/grafana/grafana/blob/main/public/app/plugins/datasource/elasticsearch/QueryBuilder.ts#L481), we are adding defaults and processing log query, this is missing on backend. 

Therefore in this PR, we are:
1. Adjusting size to be either set limit or 500
2. Processing log query as document query applying same defaults
3. Adding date histogram aggregation
4. Adding highlight params 

With this PR, processing of log query on backend and frontend should be the same. 

I have also added snapshot tests + couple of additional unit tests.

### CODE REVIEW

File: pkg/tsdb/elasticsearch/time_series_query.go [hunk 1]
```diff
@@ -78,18 +78,60 @@ func (e *timeSeriesQuery) processQuery(q *Query, ms *es.MultiSearchRequestBuilde
...
+
+		if isRawData {
+			b.Size(metric.Settings.Get("size").MustInt(500))
+			b.SortDesc(e.client.GetTimeField(), "boolean")
+			b.SortDesc("_doc", "")
+			b.AddDocValueField(e.client.GetTimeField())
+			return nil
+		}
+
+		if isLogs {
```
Comment (ivanahuckova) [inline pkg/tsdb/elasticsearch/time_series_query.go:111]: In the future, we should refactor this to separate functions, but I want to do minimal changes. When everything is passing, then refactor can be done in separate PR.

File: pkg/tsdb/elasticsearch/time_series_query.go [hunk 2]
```diff
@@ -72,23 +72,45 @@ func (e *timeSeriesQuery) processQuery(q *Query, ms *es.MultiSearchRequestBuilde
 	b.Size(0)
 	filters := b.Query().Bool().Filter()
 	filters.AddDateRangeFilter(e.client.GetTimeField(), to, from, es.DateFormatEpochMS)
-
-	if q.RawQuery != "" {
-		filters.AddQueryStringFilter(q.RawQuery, true)
-	}
+	filters.AddQueryStringFilter(q.RawQuery, true)
 
 	if len(q.BucketAggs) == 0 {
```
Comment (ivanahuckova) [inline pkg/tsdb/elasticsearch/time_series_query.go:77]: The plan is to refactor this to whole `processQuery` function, but in this PR, I wanted to do as little changes as possible, so it is easier to review. Then, when we have all tests passing, I'll work on refactor.

File: pkg/tsdb/elasticsearch/time_series_query.go [hunk 3]
```diff
@@ -72,23 +72,45 @@ func (e *timeSeriesQuery) processQuery(q *Query, ms *es.MultiSearchRequestBuilde
 	b.Size(0)
 	filters := b.Query().Bool().Filter()
 	filters.AddDateRangeFilter(e.client.GetTimeField(), to, from, es.DateFormatEpochMS)
-
-	if q.RawQuery != "" {
-		filters.AddQueryStringFilter(q.RawQuery, true)
-	}
+	filters.AddQueryStringFilter(q.RawQuery, true)
```
Comment (ivanahuckova) [inline pkg/tsdb/elasticsearch/time_series_query.go:75]: We are already handling empty string scenario in the query.

File: pkg/tsdb/elasticsearch/client/search_request.go [hunk 4]
```diff
@@ -92,6 +92,19 @@ func (b *SearchRequestBuilder) AddDocValueField(field string) *SearchRequestBuil
...
 
+// Add highlights to the search request for log queries
+func (b *SearchRequestBuilder) AddHighlight() *SearchRequestBuilder {
+	b.customProps["highlight"] = map[string]interface{}{
+		"fields": map[string]interface{}{
+			"*": map[string]interface{}{},
+		},
+		"pre_tags":      []string{"@HIGHLIGHT@"},
+		"post_tags":     []string{"@/HIGHLIGHT@"},
+		"fragment_size": 2147483647,
```
Comment (gabor) [inline pkg/tsdb/elasticsearch/client/search_request.go:103]: what an interesting number! 😄 

i looked it up, and it is the largest 32bit integer, when dealing with signed integers (meaning integers that can do negative numbers too, they go from `-2147483648` to `2147483647`)

Comment (ivanahuckova) [reply pkg/tsdb/elasticsearch/client/search_request.go:103]: Fyi https://www.agilytic.be/blog/tech-talk-elasticsearch-highlighting-kibana if you are interested in more reading on this.

**Review by gabor — APPROVED** [review]
LGTM!

i do wonder, if we should extract all these magic numbers (`500` and that large-integer etc.), and use go-constants or something maybe.

**Comment by ivanahuckova** [comment]
>i do wonder, if we should extract all these magic numbers (500 and that large-integer etc.), and use go-constants or something maybe.

Absolutely. 500 and other contants are used across bunch of places. I'll add it to TODO.

**Pull request merged**